### PR TITLE
fix(seq2seq): opt_state from ckpt + limit cache

### PR DESCRIPTION
**Pull request by borisdayma**
Changes:
* actually update opt_state
* limit artifacts cache size
* fix a bug when not setting `eval_steps`

### CODE REVIEW

**Comment by borisdayma** [comment]
I think it can't hurt.
I'll be able to test it when I can set up the TPU with more storage.
If it works, we should not have the loss going back up initially.

**Comment by pcuenca** [comment]
That's interesting!

File: dev/seq2seq/run_seq2seq_flax.py [hunk 1]
```diff
@@ -442,6 +438,7 @@ def restore_state(state, artifact_dir):
         if (Path(artifact_dir) / 'opt_state.msgpack').exists():
             with (Path(artifact_dir) /  'opt_state.msgpack').open('rb') as f:
                 opt_state = from_bytes(state.opt_state, f.read())
+            state.replace(opt_state=opt_state)
```
Comment (patil-suraj) [inline dev/seq2seq/run_seq2seq_flax.py:441]: Great catch!

File: dev/seq2seq/run_seq2seq_flax.py [hunk 2]
```diff
@@ -836,6 +833,10 @@ def run_save_model(state, step, epoch, eval_metrics=None):
                 artifact.add_file(str(Path(training_args.output_dir) / 'training_state.json'))
                 wandb.run.log_artifact(artifact)
 
+                # save some space
+                c = wandb.wandb_sdk.wandb_artifacts.get_artifacts_cache()
+                c.cleanup(wandb.util.from_human_size("15GB"))
```
Comment (patil-suraj) [inline dev/seq2seq/run_seq2seq_flax.py:838]: Nice!

File: dev/seq2seq/run_seq2seq_flax.py [hunk 3]
```diff
@@ -866,7 +867,7 @@ def run_save_model(state, step, epoch, eval_metrics=None):
                 # log metrics
                 wandb_log(unreplicate(train_metric), step=global_step, prefix='train')
 
-            if global_step % training_args.eval_steps == 0:
+            if training_args.eval_steps and global_step % training_args.eval_steps == 0:
```
Comment (patil-suraj) [inline dev/seq2seq/run_seq2seq_flax.py:870]: it might not be necessary, `eval_steps` is set to 500 by default.

**Review by patil-suraj — APPROVED** [review]
LGTM! Great catch with the `opt_state`!

File: dev/seq2seq/run_seq2seq_flax.py (hunk 3 above)
Comment (borisdayma) [reply dev/seq2seq/run_seq2seq_flax.py:870]: I had an issue with it. Looks like it's None: https://huggingface.co/transformers/main_classes/trainer.html#trainingarguments

Comment (patil-suraj) [reply dev/seq2seq/run_seq2seq_flax.py:870]: yes, it's `None`, when it's `None` it's set to `logging_steps` which is 500 by default. You could find this info in the docstring.

Comment (borisdayma) [reply dev/seq2seq/run_seq2seq_flax.py:870]: Apparently it works only if you set `evaluation_strategy` to `steps`, otherwise it stays to `None`

**Comment by borisdayma** [comment]
So I'm not convinced it worked:

image

I would expect the eval_loss to go down immediately.
I wonder if there's some init of the optimizer later happening which sets random values and overwrite these.

**Comment by borisdayma** [comment]
Getting a little closer.

The error I have now is:
```
Traceback (most recent call last):
  File "run_seq2seq_flax.py", line 953, in <module>
    main()
  File "run_seq2seq_flax.py", line 891, in main
    state, train_metric = p_train_step(state, batch)
  File "run_seq2seq_flax.py", line 728, in train_step
    new_state = jax.lax.cond(
TypeError: true_fun and false_fun output must have same type structure, got PyTreeDef(xxxxxxx)
and PyTreeDef(xxxxxxxx)
```

The content is very long. I compared it and the difference is that one definition uses `optax._src.factorized.FactoredState` while the other uses `flax.serialization.FactoredState` (it is used at 208 places in each PyTreeDef).

The same error happens if I don't use `state.replace` but just use:
```
state = TrainState(
    apply_fn=model.__call__,
    params=model.params,
    tx=optimizer,
    dropout_rng=dropout_rng,
    grad_accum=jax.tree_map(jnp.zeros_like, model.params),
    optimizer_step=optimizer_step,
    step=step,
    opt_state=opt_state
)
```

I may try `flax.training.checkpoints.save_checkpoint` and `flax.training.checkpoints.restore_checkpoint` instead (as used in [flax imagenet example](https://github.com/google/flax/blob/main/examples/imagenet/train.py))

File: dev/seq2seq/run_seq2seq_flax.py [hunk 4]
```diff
@@ -668,7 +666,8 @@ def decay_mask_fn(params):
     )
     if model_args.from_checkpoint is not None:
         # restore optimizer state, step and optimizer_step
-        restore_state(state, artifact_dir)
+        step, optimizer_step, opt_state = restore_state(state, artifact_dir)
+        state = state.replace(step=step, optimizer_step=optimizer_step, opt_state=opt_state)
```
Comment (pcuenca) [inline dev/seq2seq/run_seq2seq_flax.py:670]: Looks good!

**Review by pcuenca — APPROVED** [review]

**Comment by pcuenca** [comment]
> I may try `flax.training.checkpoints.save_checkpoint` and `flax.training.checkpoints.restore_checkpoint` instead (as used in [flax imagenet example](https://github.com/google/flax/blob/main/examples/imagenet/train.py))

I think those functions may be intended for tensorflow checkpoints.

I tried to use `flax.serialization.msgpack_restore` and `flax.serialization.from_state_dict` but they resulted in the same error. I will investigate if the types can be converted to Optax somehow.

**Comment by borisdayma** [comment]
I tested it:
- streaming
- non-streaming
- resume from checkpoint

All scenarios seem to work.
Let's merge it if nobody has any comment.
Then I'll work on simplifying gradient accumulation / model definition.

**Pull request merged**